### ISSUE TITLE
Should 6.2.0 bridge work with 6.2.1 OpenMAMA?

### ISSUE THREAD

**Issue by dfedorov-solace** [OpenMAMA/OpenMAMA#281]
# Should 6.2.0 bridge work with 6.2.1 OpenMAMA?
## Should 6.2.0 bridge work with 6.2.1 OpenMAMA?

6.2.0 bridge was compiled with USE_32_BIT_TIME and 6.2.1 OpenMAMA compiled without it.
This would make the bridge essentially incompatible and according to https://openmama.github.io/openmama_versioning.html the version of 6.2.1 should really be 6.3.0.

## Areas Affected
- [X] MAMAC
- [X] MAMACPP
- [X] MAMADOTNET
- [X] MAMAJNI
- [X] MAMDA
- [X] MAMDACPP
- [X] MAMDADOTNET
- [X] MAMDAJNI
- [X] Visual Studio
- [X] SCons
- [X] Unit Tests
- [X] Examples

## Details
6.2.0 bridge was compiled with USE_32_BIT_TIME and 6.2.1 OpenMAMA compiled without it.
This would make the bridge essentially incompatible and according to https://openmama.github.io/openmama_versioning.html the version of 6.2.1 should really be 6.3.0.

Is this correct?

**Comment by fquinner** [comment]
Did you try it and it failed?

**Comment by dfedorov-solace** [comment]
Arcontech did and they see an exception thrown when calling call mamaMsg_addDateTime().
Let's see if we find out that this is a real issue when we start testing 6.2.1 GA, what do you think should happen?

**Comment by fquinner** [comment]
I think we should see if there's a problem, verify whether or not this is the cause and then consider our position rather than speculate and theorize. I've been surprised by windows many times. We also need to consider this issue is limited to 32 bit windows and if we had many users of that in the wild, I expect they would have reported problems in their applications before now which probably weren't compiled with that flag.

**Comment by dfedorov-solace** [comment]
Great, thank you I'm going to stop speculating and theorizing now, let's wait for surprises.

**Comment by nmontgomery-arcontech** [comment]
I don't know if it qualifies as a surprise, but from the Arcontech perspective I've uncovered one further interesting piece of information: this issue only occurs with Debug builds. 

I've got local Win32 Rel/Dbg builds of OpenMama 6.2.1rc1 and the Rel/Dbg binaries of the Solace bridge. I think Solace and Arcontech should follow this up separately until we can confirm whether or not it's an OpenMAMA issue.

**Comment by cjwmorgan-sol-sys** [comment]
I have been able to show that there data integrity loss when using an application compiled with 6.2.0 OpenMAMA uses OpenMAMA 6.2.1rc1 when using the `mamaDateTime_getFromStructTimeSpec` function. The size of `struct timespec` shows a sizeof __8 bytes__ (4 byte time_t sec + 4 byte long nano) when compiled against 6.2.0 and a size of __16 bytes__ (8 byte time_t + 4 byte long nano + padding) when compiled against 6.2.1rc1. As a result when a 6.2.0 application using 6.2.1rc1 the __byte 8__ timespec is passed to the 6.2.1rc1 code which has timspec defined as __16 bytes__ so it would write 8 bytes of time_t to seconds (writing over both seconds and nano seconds) and 4 bytes to nano seconds (in invalid memory from the stack). 

Here is the sample program:

```
mamaDateTime dt=NULL;
mama_i64_t seconds;
struct timespec time={0,0};

mamaDateTime_create(&dt);
seconds=0xF8000000000000C0;
mamaDateTime_setEpochTimeF64(dt,(mama_f64_t)seconds);
mamaDateTime_getStructTimeSpec(dt, &time);
printf("DateTime1 Epoch Seconds %lld:: Check TimeSpec value: Seconds: %ld, Nano Seconds: %d, Seconds hex: 0x%x, Nano Seconds hex: 0x%x\n", seconds, time.tv_sec, time.tv_nsec, time.tv_sec, time.tv_nsec);
```
Output:
```
DateTime1 Epoch Seconds -576460752303423296:: Check TimeSpec value: Seconds: 192, Nano Seconds: -134217728, Seconds hex: 0xc0, Nano Seconds hex: 0xf8000000
```

The bottom 32 bits of seconds make up time.tv_sec and the top 32 bits of seconds in time.tv_nsec since my machine store the value in little endian order.

All structures containing time_t have changed from 6.2.0 and 6.2.1rc1 so any application or bridge using API function with these structure must be recompiled using 6.2.1rc1 or they will suffer from this problem. This only occurs on the Win32 Platform since its the only platform with this change between 6.2.0 and 6.2.1rc1.

Here is a list** of affected structures on Win32 for applications and bridges:
* struct timespec
* struct timeval

Here is a list** of affect structures on Win32 for bridges only:
* mama_datetime_t

 ** There may be more affected structures. I have only confirmed timespec and mama_datetime_t.

**Comment by fquinner** [comment]
This is going to be the nI was hoping that there was some M$ runtime / compile time magic which would convert this for us in a safer way but alas this does not seem to be the case. It's still not clear to me though whether or not this is related to the issue that Arcontech had reported. Anyway:

> The size of struct timespec shows a sizeof 8 bytes (4 byte time_t sec + 4 byte long nano) when compiled against 6.2.0 and a size of 16 bytes (8 byte time_t + 4 byte long nano + padding) when compiled against 6.2.1rc1.

Note that it's not that it only effects win32 applications - it's that it only effects 32 bit windows applications / libraries which were compiled with the non-recommended `_USE_32BIT_TIME_T`. This applies to both windows applications and bridges (and I have no appetite in moving to 7.x.x for a change which is effectively a bug for a single target platform which is extremely unlikely to be hit). The change actually *fixes* issues which may have arisen if this macro *wasn't* used in the calling application or bridge (i.e. the default microsoft compiler options which would apply to the majority of users). For example the OMNM bridge never included it and neither do the tick42 bridges. Even microsoft don't recommend using it - that macro is a temporary hack which was introduced 10 years ago to get people to migrate to using 64 bit time_t values. This is effectively a hangover from that change which has outlived its purpose.

With that in mind and to answer the question, it's expected that 6.2.0 will work fine with 6.2.1 with the exception of this known issue on this single platform and only with bridges which have been compiled with this macro. Most bridges are actually not effected. The change is a bugfix for that platform - not an interface change, therefore it will be 6.2.1.

Note that the compile / runtime versions are available to the bridge during the init method though so if you want to put in your own logic to handle this edge condition, you have an interface there to let you do so.ature of the beast when making this change.

**Comment by dpauls** [comment]
Frank, I'd agree that it makes sense to consider the use of `_USE_32BIT_TIME_T` on Win32 a bug and hence not move to 7.x.x just because of this one issue on this one platform.  However, I think it makes sense to release note this so that OpenMAMA users are aware of this incompatibility.  In order for a 6.2.0 application or bridge to make use of the public `struct timespec` interface in 6.2.0, they would have had to have been compiled with `_USE_32BIT_TIME_T` defined or else it wouldn't have worked as expected.  Then, if the same bridge/application uses OpenMAMA 6.2.1 without being recompiled, it will not work as intended.  So as long as it is clear that this is a b/w incompatible change in Win32 OpenMAMA that is being made for the greater good of the future, this all sounds good to me.

Also, we were able to debug at least one exception seen by Arcontech and we validated that the root cause was due to the `struct timespec` incompatibility.  Chris created the simpler demonstration of the incompatibility for the purposes of discussion in this issue.
